### ISSUE TITLE
Enable IPv6 by default

### ISSUE THREAD

**Issue by DerEnderKeks** [blakeblackshear/frigate#12944]
**Describe what you are trying to accomplish and why in non technical terms**
I want that IPv6 is enabled by default, as IPv4 is a legacy standard, IPv6 should always be enabled/supported unless intentionally disabled.

**Describe the solution you'd like**
IPv6 should be enabled by default again, like it was in the v14 betas. The current behavoir should thus be inverted: Those that have weird setups with hosts that have IPv6 completely disabled should apply a patch to remove the v6 listening directives.

**Describe alternatives you've considered**
There are no alternatives to this. IPv6 is the most recent standard and thus should be enabled by default. Having it disabled by default just hurts the IPv6 adoption and causes additional effort for anyone with a proper setup.

**Additional context**
According to #11684, some had issues with this because of their legacy/snowflake setups with completely disabled IPv6. 
As this is a very non-standard configuration, this can't be a reason to make everyone else enable the normal behavior again. Any properly configured host isn't impacted by a listener that listens on v6, as all current OSs also listen on v4 by default. Disabling IPv6 entirely on a host is discouraged and people that do so should rightfully suffer [the consequences].

**Comment by BobTB** [comment]
Most people are not capable of managing / using IPv6, it should be optional as it is now. Never enabled by default. IPv6 is a horrible thing anyway.

**Comment by DerEnderKeks** [comment]
@BobTB you don't seem to understand IPv6, as it's literally better in basically every way over legacy IPv4, so I'm not going to take your remark very seriously, but I'll try to explain it anyway: 

The default behavior of IPv6 [sockets] makes it optional anyway. If your network doesn't have an IPv6 enabled router, you won't notice it in any way and can keep using your legacy v4-only network like it's 1990 just fine. Applications having support for IPv6 is only a problem when you intentionally cripple your system to have IPv6 completely disabled, as that breaks the dual stack sockets, which primarily listen on v6 but also automatically on v4.

**Comment by davidebeatrici** [comment]
I agree that IPv6 should be enabled by default. I found out it wasn't because Frigate couldn't connect to the camera via its IPv6 address.

**Comment by davidebeatrici** [comment]
Please note that the file to edit is `/usr/local/nginx/conf/listen.conf` now.

**Comment by doppelgrau** [comment]
I can only second that request, with the sparse documentation getting it reachable over network (VPN is v6 only) ruined now the second evening...

**Comment by scyto** [comment]
I have a PR ready to go for this if i am allowed to submit one, already tested by modifying my local listen.conf with the same settings,  i need this for the downsteam truenas apps version of the image, they quite rightly told me to fix it upstream if i wanted this :-)

here is the sample fix

https://github.com/scyto/frigate/blob/dev/docker/main/rootfs/usr/local/nginx/templates/listen.gotmpl 

Is this something i can do a PR for?

As for why this is needed - i run a dual stack IPv4/IPv6 and cllient devices will connect to IPv4 or IPv6 with some unreliability based on how they prioritize the records returned from DNS for the host node running frigate, i was getting a lot of unreliability with the home assistant hacs integration for frigate until i made this fix - home assitant has a tendency to connect to IPv6 most of the time - this is correct behaviour in a dual stack system, and no using just IPv4 address is not a good workaround as i intentionally use DNS to find IPs and have my DNS well managed to include both IPv4 and IPv6 records of all hosts

i am likely not a good enough coder to write a flag so this listen change could be toggled by say an env var... i can try if you really want me to do vibe coding /s :-). 

There should be ZERO issues listening on both IPv4 and IPv6 wether the user has IPv6 not configured, partially configured or fully configured - if the connecting device is using IPv4 nginx will only ever reply on IPv4 (nginx configuration i do know), there are zero downsides in my opinion to enabling this listen as default but happy if its done with a toggle instead

**Comment by NickM-27** [comment]
It was originally enabled by default in 0.14 and we had a significant number of support requests during the beta period, so it will definitely need to be something that is easy to toggle via config or env var, otherwise it will just be the same / opposite problem.

**Comment by scyto** [comment]
> It was originally enabled by default in 0.14 and we had a significant number of support requests during the beta period, so it will definitely need to be something that is easy to toggle via config or env var, otherwise it will just be the same / opposite problem.

thanks for the history

were there actual issues or were people having 'knee-jerk-i-don't-understand-ipv6' reactions to the idea IPv6 was 'enabled' - like BobTB above?  if the former that would imply there was something else going on, because literally the proposed change would be 100% transparent to any one with IPv6 disabled... and would actually be a good thing for anyone who is using link-local or GUAs....  

in the original implementation was IPv6 enabled frigate wide (eg in core) ? and that caused some issue?  

i am only proposing IPv6 on the nginx proxy used for web and api - not anywhere else - thats the root of why i say it literally would be invisible changes to all non IPv6 users

**Comment by NickM-27** [comment]
It was stopping Frigate from starting:

```
2024-06-01 21:43:16.798162323  [INFO] Starting NGINX...
2024-06-01 21:43:16.825700420  2024/06/01 23:43:16 [emerg] 4457#4457: socket() [::]:80 failed (97: Address family not supported by protocol)
2024-06-01 21:43:16.825705488  nginx: [emerg] socket() [::]:80 failed (97: Address family not supported by protocol)
2024-06-01 21:43:16.831632549  [INFO] Service NGINX exited with code 1 (by signal 0)
```

**Comment by scyto** [comment]
thanks, interesting, thats different from my experience of other nginx implementations where i never hit that.... hmm need to think about if thats me or nginx. --edit-- oh i see they were using the traversty that is docker-desktop on windows -its always been very incompatible with general docker networking....  i note as for Jun 25th it now supports IPv6 quite happily, and IPv6 is defaulted to on in windows for at least a decade at this point.....

but how silly nginx treats that as an emergency and not a warn..... will have to bring that up at my next D&D session...

I think i know what will fix it, i will test in a VM and see if i am right or have too much hubris....

--edit-- 
i have yet to reproduce this behaviour on linux with IPv6 hard disabled using `sysctl -w net.ipv6.conf.all.disable_ipv6=1` will keep trying

**Comment by scyto** [comment]
@NickM-27 

Ok a few things:

- the only way to generate the posted error you showed me on debian linux is to disable IPv6 on the kernel command line (interface lo always is bound to IPv6 and using sysctls or configuring interfaces to not have IPv6 didn't cause the break), i assume this is the case on debian downstream distros too, but would need to check
- the original OP error was on docker desktop for windows - that seems to have IPv6 how, but i will test
- i hear you on the need for it to be a setting and disabled by default, so.....

Please go easy on me, i am not a developer (like at all), apologies if this looks horrific, yes i used 'vibe coding' and i learnt alot about python includes. How the frigate config.py works etc

- So here is my PoC, i overloaded get_tls_settings.py to make this easier on a non-coder like me
- I added a ipv6.py in the config dir too so that i could import that in the yaml validator 'stuff' in config.py
- i am sure the listen.gotmpl stuff looks horrible to you, best i could do with combo of copilot and chatgpt - they really didn't want me to use the shortcut 'if not' approach the team used.... oh and of course eveything defaults to ipv6 off :-)

https://github.com/blakeblackshear/frigate/commit/3455cc580f1af788c493017826f3a7b9f8b054c1 (this is a compare of current code and my repo, i think, as i said this is amostly all new to me, i have done some github PRs before for other projects but still find it quiet unintuative)

tl;dr i have tested this in the following scenarios:

1. a host VM with IPv6 uttlery disabled in the kernel
2. a host VM with IPv4 and IPv6 enabled
3. tested all 4 combinations of TLS / IPv6 on and off

the only hard break  I could find to a users config is if they are #1 and set ipv6 enable true in the config.yaml

non-goals of this poc : this is not supposed to enable anything else for IPv6 in frigate - just the stuff served by nginx

next step for me is to see how i can test on my running system - i am not sure i can do a full build from source with what i did (will try at the weekend or next week) i will likely just bind mount the individual modified code files into the container.....

**I am not sure what i am asking other than:**

**- is this on the right path?**
**- would the team prefer me to create a get_ipv6_settings.py file instead on overloading the get_tls one?**
**- I assume there would need to be some unit tests, code linting and formatting tests - i am at a loss as to where they are so ask for a pointer if you would be willing?**


ps phew that was a hard 7 hours, lol ;-)

**Comment by scyto** [comment]
> - would the team prefer me to create a get_ipv6_settings.py file instead on overloading the get_tls one?

oh i see why that wouldn't work, it renders the whole file, i guess renaming the existing file to something like get_listener_settings.py would be the correct approach and updating the references in the s6-rc files

**Comment by NickM-27** [comment]
That seems reasonable to me, and we don't have any tests for this so no worries there. 

Once a PR is up we can give a more thorough review but to my eyes the approach seems fine other than I'd make the config section be called network instead of just ipv6

**Comment by scyto** [comment]
> That seems reasonable to me, and we don't have any tests for this so no worries there.
> 
> Once a PR is up we can give a more thorough review but to my eyes the approach seems fine other than I'd make the config section be called network instead of just ipv6

wow, thanks, will do, thanks for you patience too

ok, i will make the change you suggest, do you care about the filename of get_tls_settings.py being renamed to say get_network_settings.py vs not?

**Comment by NickM-27** [comment]
Something more generic than tls would likely be good

**Comment by scyto** [comment]
@NickM-27  i have a set of changes ready, wanted to test them as whole if possible. ran the CI workflow but failed on all container builds with `Error: buildx failed with: ERROR: failed to build: failed to solve: process "/bin/sh -c /build_pysqlite3.sh" did not complete successfully: exit code: 2` should i use one of the other flows to kick off, or is this a set of changes you will be happy to review if i push without me completeing a full image build?

https://github.com/blakeblackshear/frigate/compare/dev...scyto:frigate:dev

**Comment by NickM-27** [comment]
sounds like you need to sync your dev with the latest which fixed that issue

**Comment by scyto** [comment]
@doppelgrau @davidebeatrici @DerEnderKeks 
if you are interested in testing this change to see if it works, you can use an image from my branch
https://github.com/scyto/frigate/pkgs/container/frigate/490160323?tag=3bdb286-amd64 

you need to add the following to any config.yaml

```
network:
  ipv6:
    enabled: true
```

of course please do not use this image on your existing production installs / instances, use on fresh test

**Comment by scyto** [comment]
nice, looks like go2rtc is already dual stack, no other changes needed

```
# root@debian-no-ip-v6:~ss -lntup | egrep ':(1984|8554|8555|8971|5000)\b'b'
udp   UNCONN 0      0                              192.168.1.163:8555      0.0.0.0:*    users:(("go2rtc",pid=462943,fd=7))                                                                                                       
udp   UNCONN 0      0      [2600:dead:beef:7654:1c6b:3d72:8b78:4607]:8555     [::]:*    users:(("go2rtc",pid=462943,fd=8))                                                                                                       
tcp   LISTEN 0      511                                  0.0.0.0:8971      0.0.0.0:*    users:(("nginx",pid=463047,fd=7),("nginx",pid=463046,fd=7),("nginx",pid=463045,fd=7),("nginx",pid=463044,fd=7),("nginx",pid=462976,fd=7))
tcp   LISTEN 0      511                                  0.0.0.0:5000      0.0.0.0:*    users:(("nginx",pid=463047,fd=5),("nginx",pid=463046,fd=5),("nginx",pid=463045,fd=5),("nginx",pid=463044,fd=5),("nginx",pid=462976,fd=5))
tcp   LISTEN 0      511                                     [::]:8971         [::]:*    users:(("nginx",pid=463047,fd=8),("nginx",pid=463046,fd=8),("nginx",pid=463045,fd=8),("nginx",pid=463044,fd=8),("nginx",pid=462976,fd=8))
tcp   LISTEN 0      511                                     [::]:5000         [::]:*    users:(("nginx",pid=463047,fd=6),("nginx",pid=463046,fd=6),("nginx",pid=463045,fd=6),("nginx",pid=463044,fd=6),("nginx",pid=462976,fd=6))
tcp   LISTEN 0      4096                                       *:1984            *:*    users:(("go2rtc",pid=462943,fd=9))                                                                                                       
tcp   LISTEN 0      4096                                       *:8554            *:*    users:(("go2rtc",pid=462943,fd=5))                                                                                                       
tcp   LISTEN 0      4096                                       *:8555            *:*    users:(("go2rtc",pid=462943,fd=6))     
```

**Comment by maxlyth** [comment]
It looks like the IPv6 config options fail in 16.4 as the config parser throws a validation error:
```
Key     : network
Value   : {'ipv6': {'enabled': True}} 
Message : Extra inputs are not permitted
```
I can see code has been added to parse this config key and a GoTemplate has been written to use those entries for NGinX conf but the code is not imported in `frigate/config/__init__.py` and the schema parsing fails.

I am migrating to IPv6 and have HA and go2rtc running well in separate containers for WebRTC and dashboards but I'm struggling with Frigate and therefore relying on a reverse proxy (Caddy) for IPv6

**Comment by NickM-27** [comment]
This feature was implemented in 0.17

**Comment by dlitz** [comment]
For future reference, the configuration you need for 0.17.0 is:

```yaml
networking:
  ipv6:
    enabled: true
```

Note "networking" is used rather than "network".
